Storage/SQL/SQLiteHistory.swift

```diff
@@ -425,14 +441,48 @@ extension SQLiteHistory: BrowserHistory {
                       " historyID, url, title, guid, domain_id, domain" +
                       ", localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount" +
                       ", iconID, iconURL, iconDate, iconType, iconWidth" +
-                      " FROM (\(historySQL)) LEFT OUTER JOIN " +
-                      "view_history_id_favicon ON historyID = view_history_id_favicon.id"
+                      " FROM (\(historySQL)) LEFT OUTER JOIN" +
+                      " view_history_id_favicon ON historyID = view_history_id_favicon.id"
             let factory = SQLiteHistory.iconHistoryColumnFactory
-            return db.runQuery(sql, args: args, factory: factory)
+            return db.runQuery(sql, args: args, factory: factory) >>== { cursor in
+                if includeBookmarkState {
+                    // get list of tuples of ids & urls
+                    return self.addBookmarkedState(toSites: cursor.asArray())
+                }
+                return deferMaybe(cursor)
+            }
         }
 
         let factory = SQLiteHistory.basicHistoryColumnFactory
-        return db.runQuery(historySQL, args: args, factory: factory)
+        return db.runQuery(historySQL, args: args, factory: factory) >>== { cursor in
+            if includeBookmarkState {
+                // get list of tuples of ids & urls
+                return self.addBookmarkedState(toSites: cursor.asArray())
+            }
+            return deferMaybe(cursor)
+        }
+    }
+
+    func addBookmarkedState(toSites sites: [Site]) -> Deferred<Maybe<Cursor<Site>>> {
+        let historyRowIds = optFilter(sites.map { String($0.id!) })
+        let historySQL = historyRowIds.joinWithSeparator(", ")
+        let bookmarkSQL = "SELECT  \(TableHistory).id AS historyID, count(\(TableBookmarks).id) AS bookmarkCount " +
+            "FROM \(TableHistory) LEFT JOIN \(TableBookmarks) ON \(TableHistory).url = \(TableBookmarks).url " +
+            "WHERE \(TableHistory).id IN (\(historySQL)) " +
+            "GROUP BY \(TableHistory).id"
+        func BookmarkFactory(row: SDRow) -> Site {
+            // we should never fail this test as we will get a response containing only those rows that exist in the sites array, but optionals will be optionals
+            guard let siteID = row["historyID"] as? Int, let index = sites.indexOf({ $0.id == siteID })
+                else {
+                    log.error("Returning an empty site from Bookmarks. This shouldn't ever happen!");
+                    return Site(url: "", title: "")
+            }
+            let site = sites[index]
+            site.bookmarked = row.getBoolean("bookmarkCount")
```

Comment:
Ooh, relies on implementation detail of `getBoolean`! Let's make this explicitly `> 0`, please.